### PR TITLE
completionItem/resolve support

### LINKED ISSUE #616: How to contribute to `completionItem/resolve` ?
For Maven extension, we'd like to offer completion item documentation asynchronously (as it requires to read the pom which we cannot do for all proposals or completion would take to long).
Does the extension provide a way to "resolve" a completion item? I couldn't find an appropriate method.

### PR DESCRIPTION
**Pull request by datho7561**
Collect model documentation for completion items in the resolve step

Fixes #616, redhat-developer/vscode-xml#679

Signed-off-by: David Thompson <davthomp@redhat.com>

### CODE REVIEW

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/ContentModelCompletionParticipant.java [hunk 1]
```diff
@@ -352,15 +372,29 @@ private void fillAttributeValuesWithCMAttributeDeclarations(CMElementDeclaration
...
 				item.setFilterText(insertText);
 				item.setTextEdit(Either.forLeft(new TextEdit(fullRange, insertText)));
-				MarkupContent documentation = XMLGenerator.createMarkupContent(cmAttribute, value, cmElement, request);
-				item.setDocumentation(documentation);
+
+				if (request.isResolveDocumentationSupported()) {
+					DOMDocument document = request.getNode().getOwnerDocument();
+					JsonObject data = DataEntryField.createData(document.getDocumentURI(),
+							AttributeValueCompletionResolver.PARTICIPANT_ID);
+					try {
```
Comment (angelozerr) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/ContentModelCompletionParticipant.java:380]: I have 2 questions:

 * why can you not store the offset as int?
 * why you convert again the position to offset although you have it with request.getOffset();

In other words I would write:

data.addProperty(AttributeValueCompletionResolver.OFFSET_KEY, request.getOffset());

Comment (datho7561) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/ContentModelCompletionParticipant.java:380]: Thanks for the pointers! This looks like a much better way to do this

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/ContentModelCompletionParticipant.java [hunk 2]
```diff
@@ -352,15 +372,29 @@ private void fillAttributeValuesWithCMAttributeDeclarations(CMElementDeclaration
 				item.setKind(CompletionItemKind.Value);
 				item.setFilterText(insertText);
 				item.setTextEdit(Either.forLeft(new TextEdit(fullRange, insertText)));
-				MarkupContent documentation = XMLGenerator.createMarkupContent(cmAttribute, value, cmElement, request);
-				item.setDocumentation(documentation);
+
+				if (request.isResolveDocumentationSupported()) {
```
Comment (angelozerr) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/ContentModelCompletionParticipant.java:376]: I think texedit should be done in resolve process to minimize the json response size when thére are a of of items to return

Comment (datho7561) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/ContentModelCompletionParticipant.java:376]: I could look into adding the text edits to `additionalTextEdit` and send a blank `textEdit` for the unresolved completion item, but I don't know if its worthwhile, and the documentation seems to discourage this.

Comment (angelozerr) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/ContentModelCompletionParticipant.java:383]: Thanks for your investigation. Let's keep like you did with textEdit fill.

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/ContentModelCompletionParticipant.java [hunk 3]
```diff
@@ -45,14 +51,26 @@
...
 
+import com.google.gson.JsonObject;
+
 /**
  * Extension to support XML completion based on content model (XML Schema
  * completion, etc)
  */
 public class ContentModelCompletionParticipant extends CompletionParticipantAdapter {
 
+	private final HashMap<String, ICompletionItemResolveParticipant> completionResolvers;
```
Comment (angelozerr) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/ContentModelCompletionParticipant.java:62]: private final Map

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/completion/AbstractCMCompletionResolver.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+import org.eclipse.lemminx.dom.DOMAttr;
+import org.eclipse.lemminx.dom.DOMDocument;
+import org.eclipse.lemminx.dom.DOMElement;
+import org.eclipse.lemminx.dom.DOMNode;
+import org.eclipse.lemminx.services.extensions.completion.ICompletionItemResolveParticipant;
+import org.eclipse.lemminx.services.extensions.completion.ICompletionItemResolverRequest;
+import org.eclipse.lsp4j.CompletionItem;
+import org.eclipse.lsp4j.jsonrpc.CancelChecker;
+
+public abstract class AbstractCMCompletionResolver implements ICompletionItemResolveParticipant {
```
Comment (angelozerr) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/completion/AbstractCMCompletionResolver.java:12]: Add Javdoc

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/completion/AbstractCMCompletionResolver.java [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+	/**
+	 * Add the documentation from the xml model to the given completion request.
+	 *
+	 * @param request       the completion resolve request
+	 * @param toResolve     the unresolved completion item to resolve
+	 * @param parentElement the parent element to where completion was opened on
+	 * @param attr          the attribute that completion was opened on, may be null
+	 */
+	abstract void addDocumentationToCompletion(ICompletionItemResolverRequest request,
```
Comment (angelozerr) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/completion/AbstractCMCompletionResolver.java:47]: protect abstract

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/completion/AttributeNameCompletionResolver.java [hunk 6]
```diff
@@ -0,0 +1,54 @@
+package org.eclipse.lemminx.extensions.contentmodel.participants.completion;
```
Comment (angelozerr) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/completion/AttributeNameCompletionResolver.java:1]: Add header

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/completion/AttributeNameCompletionResolver.java [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+import org.eclipse.lemminx.extensions.contentmodel.model.CMDocument;
+import org.eclipse.lemminx.extensions.contentmodel.model.CMElementDeclaration;
+import org.eclipse.lemminx.extensions.contentmodel.model.ContentModelManager;
+import org.eclipse.lemminx.extensions.contentmodel.utils.XMLGenerator;
+import org.eclipse.lemminx.services.extensions.completion.ICompletionItemResolverRequest;
+import org.eclipse.lemminx.uriresolver.CacheResourceDownloadingException;
+import org.eclipse.lsp4j.CompletionItem;
+import org.eclipse.lsp4j.MarkupContent;
+
+public class AttributeNameCompletionResolver extends AbstractCMCompletionResolver {
```
Comment (angelozerr) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/completion/AttributeNameCompletionResolver.java:17]: Add Javdoc

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/completion/AttributeNameCompletionResolver.java [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+						return;
+					}
+				}
+			}
+		} catch (CacheResourceDownloadingException e) {
+		}
+
+	}
+
+	private MarkupContent getDocumentationForAttributeValue(CMElementDeclaration cmElement, String attributeName,
```
Comment (angelozerr) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/completion/AttributeNameCompletionResolver.java:44]: private static

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/completion/AttributeValueCompletionResolver.java [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+						toResolve.setDocumentation(documentation);
+						return;
+					}
+				}
+			}
+		} catch (CacheResourceDownloadingException e) {
+		}
+	}
+
+	private MarkupContent getDocumentationForAttributeValue(CMElementDeclaration cmElement, String attributeName,
```
Comment (angelozerr) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/completion/AttributeValueCompletionResolver.java:63]: private static

**Comment by angelozerr** [comment]
@mickaelistria pay attention the package name for ICompletionParticipant will change.

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/ResolveCompletionItemRequest.java [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+import org.eclipse.lemminx.dom.DOMDocument;
+import org.eclipse.lemminx.services.data.DataEntryField;
+import org.eclipse.lemminx.services.extensions.IComponentProvider;
+import org.eclipse.lemminx.services.extensions.completion.BaseCompletionRequest;
+import org.eclipse.lemminx.services.extensions.completion.ICompletionItemResolverRequest;
+import org.eclipse.lemminx.settings.SharedSettings;
+import org.eclipse.lemminx.settings.XMLCompletionSettings;
+import org.eclipse.lsp4j.CompletionItem;
+
+public class ResolveCompletionItemRequest extends BaseCompletionRequest implements ICompletionItemResolverRequest {
```
Comment (angelozerr) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/ResolveCompletionItemRequest.java:23]: Add Javdoc

**Review by angelozerr — CHANGES_REQUESTED** [review]

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/XMLCompletions.java [hunk 11]
```diff
@@ -282,6 +286,32 @@ public CompletionList doComplete(DOMDocument xmlDocument, Position position, Sha
 		}
 	}
 
+	public CompletionItem resolveCompletionItem(CompletionItem unresolved, DOMDocument xmlDocument,
```
Comment (angelozerr) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/XMLCompletions.java:289]: Add Javdoc

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/extensions/completion/BaseCompletionRequest.java [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+* Contributors:
+*     Red Hat Inc. - initial API and implementation
+*******************************************************************************/
+package org.eclipse.lemminx.services.extensions.completion;
+
+import org.eclipse.lemminx.dom.DOMDocument;
+import org.eclipse.lemminx.services.extensions.IComponentProvider;
+import org.eclipse.lemminx.settings.SharedSettings;
+
+public class BaseCompletionRequest implements IBaseCompletionRequest {
```
Comment (angelozerr) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/extensions/completion/BaseCompletionRequest.java:18]: Add Javdoc

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/extensions/completion/IBaseCompletionRequest.java [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+* Contributors:
+*     Red Hat Inc. - initial API and implementation
+*******************************************************************************/
+package org.eclipse.lemminx.services.extensions.completion;
+
+import org.eclipse.lemminx.dom.DOMDocument;
+import org.eclipse.lemminx.services.extensions.IComponentProvider;
+import org.eclipse.lemminx.settings.SharedSettings;
+
+public interface IBaseCompletionRequest extends IComponentProvider {
```
Comment (angelozerr) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/extensions/completion/IBaseCompletionRequest.java:18]: Add Javdoc

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/extensions/completion/ICompletionItemResolveParticipant.java [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+*     Red Hat Inc. - initial API and implementation
+*******************************************************************************/
+package org.eclipse.lemminx.services.extensions.completion;
+
+import org.eclipse.lsp4j.CompletionItem;
+import org.eclipse.lsp4j.jsonrpc.CancelChecker;
+
+public interface ICompletionItemResolveParticipant {
+
+	CompletionItem resolveCompletionItem(ICompletionItemResolverRequest request, CancelChecker cancelChecker);
```
Comment (angelozerr) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/extensions/completion/ICompletionItemResolveParticipant.java:19]: Add Javdoc

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/extensions/completion/ICompletionItemResolverRequest.java [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+*
+* Contributors:
+*     Red Hat Inc. - initial API and implementation
+*******************************************************************************/
+package org.eclipse.lemminx.services.extensions.completion;
+
+import org.eclipse.lemminx.services.extensions.ISharedSettingsRequest;
+import org.eclipse.lsp4j.CompletionItem;
+
+public interface ICompletionItemResolverRequest extends IBaseCompletionRequest, ISharedSettingsRequest {
```
Comment (angelozerr) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/extensions/completion/ICompletionItemResolverRequest.java:17]: Add Javdoc and for methods too

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/extensions/completion/ICompletionParticipant.java [hunk 16]
```diff
@@ -52,4 +52,8 @@ void onAttributeValue(String valuePrefix, ICompletionRequest request, ICompletio
 	void onDTDSystemId(String valuePrefix, ICompletionRequest request, ICompletionResponse response, CancelChecker cancelChecker)
 			throws Exception;
 
+	default ICompletionItemResolveParticipant getResolveCompletionItemParticipant(String participantId) {
```
Comment (angelozerr) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/extensions/completion/ICompletionParticipant.java:55]: Add Javdoc

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/settings/XMLCompletionSettings.java [hunk 17]
```diff
@@ -92,6 +95,14 @@ public boolean isCompletionSnippetsSupported() {
 				&& completionCapabilities.getCompletionItem().getSnippetSupport();
 	}
 
+	public boolean isResolveDocumentationSupported() {
```
Comment (angelozerr) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/settings/XMLCompletionSettings.java:98]: Add Javdoc

**Comment by angelozerr** [comment]
I tried the usecase from https://github.com/redhat-developer/vscode-xml/issues/679 and it works like a charm:

CompletionResolveDemo

Great job @datho7561 !

**Pull request merged**